content/en/blog/_posts/2024-04-15-devsecops-in-action-k8s-secrets.md

```diff
@@ -0,0 +1,535 @@
+	}
+}
+```
+In fact everything works and you can stop here. Unfortunately in the real world
+there are a lot of already working services and they must have implementation of
+additional environment variables. This isn't always possible and this approach
+can have specific limitations. In order to reduce the amount of limitations it's
+necessary to implement an approach based on support of both
+functions (variables) but without additional environment variables.
+This approach allows not to implement additional variable but to use the current
+variable for two cases:
+
+* legacy: using env variable for storing sensitive data
+* new: using env variable for storing path to file with sensitive data and
+override current value to value from the file
+
+In order to distinguish sensitive data values from path to file, by using prefix.
+This will allow us to use the same variable and follow OWASP recommendations.
+
+First let's define variables and constants:
+
+*[cmd/main.go](https://github.com/efrikin/devsecops-in-action-kubernetes-secrets/blob/8c8177b4955ce9cd9686c2746e5e2a810b9c2f62/cmd/main.go#L12-L20)*
+```go
+const (
+	envVar = "DEMO_SECRET__PASSWD"
+	filePrefix = "file://"
+)
+
+var (
+	envVal, filePath string
+	isExistFile, isExistEnv, changed bool
+)
+```
+
+Next, it's necessary to change the `fileExist` function. It's supposed to return
+a path to the file without a prefix if the file exists.
+
+*[cmd/main.go](https://github.com/efrikin/devsecops-in-action-kubernetes-secrets/blob/8c8177b4955ce9cd9686c2746e5e2a810b9c2f62/cmd/main.go#L31-L43)*
+```go
+func fileExist(path string) (string, bool) {
+	if strings.HasPrefix(path, filePrefix) {
+		path = strings.TrimPrefix(path, filePrefix)
+		info, err := os.Stat(path)
+		if err == nil {
+			return path, !info.IsDir()
+		}
+		if errors.Is(err, os.ErrNotExist) {
+			return "", false
+		}
+	}
+	return "", false
+}
+```
+
+Magic happens in the `httpHandle` function. Environment variable value is passed
+into function, if it's a file it will be read and current environment variable
+value will be overridden by a new value that was read from the file.
+
+*[cmd/main.go](https://github.com/efrikin/devsecops-in-action-kubernetes-secrets/blob/8c8177b4955ce9cd9686c2746e5e2a810b9c2f62/cmd/main.go#L45-L59)*
+```go
+func httpHandle(w http.ResponseWriter, r *http.Request) {
+	if !changed {
+		if envVal, isExistEnv = os.LookupEnv(envVar); isExistEnv {
+			if filePath, isExistFile = fileExist(envVal); isExistFile {
+				body, err := os.ReadFile(filePath)
+				if err != nil {
+					log.Fatalf("unable to read file: %v", err)
+				}
+				os.Setenv(envVar, string(body))
+				envVal = os.Getenv(envVar)
+				changed = true
+			}
+		}
+	}
+}
+```
+
+Finally, if define `DEMO_SECRET__PASSWD=file:///secrets/secret.file`, deploy
+application and make request to `root` location, response will contain file
+content which is defined in the environment variable:
+
+```bash
+curl -v localhost:8080
+...
```

Comment:
Ideally, [separate commands from output](https://kubernetes.io/docs/contribute/style/style-guide/#separate-commands-from-output).

Reply:
Thank you for your suggest

Fixed.